Do not migrate legacy classes with custom values

RobinMalfait opened the pull request:
This PR fixes an issue where we migrated classes such as `rounded` to `rounded-sm` (see: https://github.com/tailwindlabs/tailwindcss/pull/14875)

However, if you override the values in your `tailwind.config.js` file, then the migration might not be correct.

This PR makes sure to only migrate the classes if you haven't overridden the values in your `tailwind.config.js` file.

RobinMalfait committed:
create git repo for integration tests

This allows us to commit the "before" state, once you are debugging you
can see the after state as a diff.

RobinMalfait committed:
add failing test

RobinMalfait committed:
do not migrate legacy classes if custom config is used

RobinMalfait commented:
Initial implementation: just check whether you have `theme[themekey]` or `theme.extend[themekey]` defined. An improvement to this is actually comparing the value from the "before" state to the value of the "after" state. if those are the same, then we can safely migrate the value.

RobinMalfait committed:
make migrations async

RobinMalfait committed:
only migrate suffix-less if it's safe to do so

RobinMalfait committed:
split `simple legacy classes` into a `legacy classes` one

The simple legacy classes migration isn't as simple anymore now that we
have to check the design system. Split it into 2 separate migrations
instead.

RobinMalfait committed:
update changelog

RobinMalfait commented on integrations/utils.ts:
```diff
@@ -421,6 +421,20 @@ export function test(
 
       options.onTestFinished(dispose)
 
+      // Make it a git repository, and commit all files
+      if (only || debug) {
+        try {
+          await context.exec('git init', { cwd: root })
+          await context.exec('git add --all', { cwd: root })
+          await context.exec('git commit -m "before migration"', { cwd: root })
+        } catch (error: any) {
+          console.error(error)
+          console.error(error.stdout?.toString())
+          console.error(error.stderr?.toString())
+          throw error
+        }
+      }
+
```
Did this to improve the logs in `.only` mode, and an easier before/after state in `.debug` mode.

RobinMalfait commented on CHANGELOG.md:
```diff
@@ -7,7 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-- Nothing yet!
+### Fixed
+
+- _Upgrade (experimental)_: Do not migrate legacy classes with custom values ([#14976](https://github.com/tailwindlabs/tailwindcss/pull/14976))
```
Maybe something like:
```suggestion
- _Upgrade (experimental)_: Do not migrate legacy classes with custom defined theme values ([#14976](https://github.com/tailwindlabs/tailwindcss/pull/14976))
```
🤔

philipp-spiess replied:
What about
 
```suggestion
- _Upgrade (experimental)_: Do not rename classes using custom defined theme values ([#14976](https://github.com/tailwindlabs/tailwindcss/pull/14976))
```

philipp-spiess commented on packages/@tailwindcss-upgrade/src/template/codemods/legacy-classes.ts:
```diff
@@ -0,0 +1,110 @@
+    SEEDED.add(designSystem)
+  }
+
+  let defaultDesignSystem = await __unstable__loadDesignSystem('@import "tailwindcss";', {
+    base: __dirname,
+  })
```
Should move this outside of the function scope, it's doing this for every candidate otherwise

philipp-spiess commented on packages/@tailwindcss-upgrade/src/template/codemods/legacy-classes.ts:
```diff
@@ -0,0 +1,110 @@
+  _userConfig: Config,
+  rawCandidate: string,
+): Promise<string> {
+  // Prepare design system with the unknown legacy classes
+  if (!SEEDED.has(designSystem)) {
+    for (let old in LEGACY_CLASS_MAP) {
+      designSystem.utilities.static(old, () => [])
+    }
+    SEEDED.add(designSystem)
+  }
```
Not needed anymore

RobinMalfait replied:
Updated the comment, the reason we do need this is to make the migration easier by registering the `shadow-sm` (or all old classes) as a static utility so that the `candidate.root` is the actual class.

If we don't do this, then we have to deal with all kinds of candidate types (which might be fine). But that also means looking at the value and migrating those.

At the end of the day it's easier to migrate:
```json
[
  {
    "kind": "static",
    "root": "shadow-sm",
    "variants": [],
    "important": false,
    "raw": "shadow-sm"
  },
]
```

Compared to this:
```json
[
  {
    "kind": "functional",
    "root": "shadow",
    "modifier": null,
    "value": {
      "kind": "named",
      "value": "sm",
      "fraction": null
    },
    "variants": [],
    "important": false,
    "raw": "shadow-sm"
  }
]
```

philipp-spiess replied:
I wonder if this might cause issues downstream though since the `designSystem` is passed through the whole migration pipeline and now it has `shadow-sm` added everywhere wrongly, just for this migration.

RobinMalfait replied:
Had an initial fix for this: https://github.com/tailwindlabs/tailwindcss/pull/14976/commits/d644a3f06be300f014fca49cadafe033526a6f35

Then improved it further: https://github.com/tailwindlabs/tailwindcss/pull/14976/commits/4bf1b7a508921e937adb0c7f9de6a40cea672700

RobinMalfait committed:
only compute the default design system once

RobinMalfait committed:
Ensure it's safe to perform suffix-less migrations (#14979)

This PR makes sure that migrations from suffix-less candidates (e.g.:
`rounded`, `blur`, `shadow`) are safe to be migrated.

In some code snippets that's not always the case.

Given the following code snippet:
```tsx
type Star = [
  x: number,
  y: number,
  dim?: boolean,
  blur?: boolean,
  rounded?: boolean,
  shadow?: boolean,
]

function Star({ point: [cx, cy, dim, blur, rounded, shadow] }: { point: Star }) {
  return <svg class="rounded shadow blur" filter={blur ? 'url(…)' : undefined} />
}
```

Without this change, it would result in:
```tsx
type Star = [
  x: number,
  y: number,
  dim?: boolean,
  blur-sm?: boolean,
  rounded-sm?: boolean,
  shadow-sm?: boolean,
]

function Star({ point: [cx, cy, dim, blur-sm, rounded-sm, shadow-sm] }: { point: Star }) {
  return <svg class="rounded-sm shadow-sm blur-sm" filter={blur-sm ? 'url(…)' : undefined} />
}
```

But with this change, it results in:
```tsx
type Star = [
  x: number,
  y: number,
  dim?: boolean,
  blur?: boolean,
  rounded?: boolean,
  shadow?: boolean,
]

function Star({ point: [cx, cy, dim, blur, rounded, shadow] }: { point: Star }) {
  return <svg class="rounded-sm shadow-sm blur-sm" filter={blur ? 'url(…)' : undefined} />
}
```

Notice how the classes inside the `class` attribute _are_ converted, but
the ones in the types or as part of the JavaScript code (e.g.:
`filter={blur ? 'url(…)' : undefined}`) are not.

---------

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
Update CHANGELOG.md

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
replace candidates based on type

RobinMalfait committed:
refactor: apply proper migrations on existing utilities in the system

Instead of creating an ad-hoc static utility for the candidate, we
simplify the candidate to it's base form, then perform the change and
then go back.

E.g.:

1. Incoming string of `hover:blur!`
2. Is turned into the candidate AST for `hover:blur!`
3. Is simplified such that variants and important is stripped
4. Is printed as a string `blur`
5. Find replacement `blur-sm`
6. Parse into candidate AST
7. Re-apply the variants and important
8. Stringified `hover:blur-sm!`
